File: plugins/typescript/src/serve.test.ts
// SPDX-License-Identifier: MIT

import assert from "node:assert/strict";
import { Readable, Writable } from "node:stream";
import test from "node:test";
import { parseFrame, PROTOCOL, type Frame } from "./protocol.js";
import { serve, type Analyzer } from "./serve.js";

/** Collects everything written, so a test can assert on frames rather than text. */
class Sink extends Writable {
  chunks: string[] = [];
  override _write(chunk: Buffer, _enc: string, cb: () => void): void {
    this.chunks.push(chunk.toString());
    cb();
  }
  text(): string {
    return this.chunks.join("");
  }
  frames(): Frame[] {
    return this.text()
      .split("\n")
      .filter((l) => l.trim() !== "")
      .map((l) => {
        const parsed = parseFrame(l);
        assert.ok("frame" in parsed, `not a frame: ${l}`);
        return parsed.frame;
      });
  }
}

function converse(analyzers: Analyzer[], ...lines: string[]): Promise<{ out: Sink; log: Sink }> {
  const out = new Sink();
  const log = new Sink();
  const input = Readable.from([lines.join("\n") + "\n"]);
  return serve({ name: "test", version: "0.0.1", analyzers, input, output: out, log }).then(() => ({ out, log }));
}

const noop: Analyzer = {
  id: "demo",
  order: 10,
  lane: "deterministic",
  async run() {
    return { findings: [] };
  },
};

test("full conversation", async () => {
  const { out } = await converse(
    [noop],
    JSON.stringify({ type: "hello", protocol: PROTOCOL, host: "test" }),
    JSON.stringify({ type: "describe" }),
    JSON.stringify({ type: "analyze", analyzer: "demo", request: { root: "/repo", changed: [] } }),
    JSON.stringify({ type: "bye" }),
  );
  const frames = out.frames();
  assert.deepEqual(
    frames.map((f) => f.type),
    ["hello", "describe", "findings"],
  );
  assert.equal(frames[0]?.protocol, PROTOCOL);
  assert.equal(frames[0]?.plugin, "test");
  assert.deepEqual(frames[1]?.analyzers, [{ id: "demo", lane: "deterministic", order: 10, available: true }]);
});

test("refuses a protocol mismatch and says so", async () => {
  const { out } = await converse([noop], JSON.stringify({ type: "hello", protocol: 99 }));
  const frames = out.frames();
  assert.equal(frames.length, 1);
  assert.equal(frames[0]?.type, "error");
  assert.match(frames[0]?.message ?? "", /protocol 99/);
});

test("refuses work before the handshake", async () => {
  const { out } = await converse([noop], JSON.stringify({ type: "analyze", analyzer: "demo" }));
  assert.match(out.frames()[0]?.message ?? "", /before hello/);
});

test("an analyzer that throws does not end the conversation", async () => {
  const broken: Analyzer = {
    id: "broken",
    order: 20,
    lane: "deterministic",
    async run() {
      throw new Error("no configuration found");
    },
  };
  const { out, log } = await converse(
    [broken, noop],
    JSON.stringify({ type: "hello", protocol: PROTOCOL }),
    JSON.stringify({ type: "analyze", analyzer: "broken" }),
    JSON.stringify({ type: "analyze", analyzer: "demo" }),
    JSON.stringify({ type: "bye" }),
  );
  const frames = out.frames();
  assert.deepEqual(
    frames.map((f) => f.type),
    ["hello", "error", "findings"],
  );
  assert.equal(frames[1]?.analyzer, "broken");
  assert.match(log.text(), /no configuration found/);
});

test("an unavailable analyzer is described, not hidden", async () => {
  const absent: Analyzer = {
    id: "absent",
    order: 30,
    lane: "deterministic",
    async unavailable() {
      return "no eslint configuration found";
    },
    async run() {
      return { findings: [] };
    },
  };
  const { out } = await converse(
    [absent],
    JSON.stringify({ type: "hello", protocol: PROTOCOL }),
    JSON.stringify({ type: "describe" }),
    JSON.stringify({ type: "bye" }),
  );
  const [descriptor] = out.frames()[1]?.analyzers ?? [];
  assert.equal(descriptor?.available, false);
  assert.equal(descriptor?.unavailable, "no eslint configuration found");
});

test("a line that is not a frame is reported and the loop continues", async () => {
  const { out, log } = await converse(
    [noop],
    JSON.stringify({ type: "hello", protocol: PROTOCOL }),
    "Debugger listening on ws://127.0.0.1:9229",
    JSON.stringify({ type: "bye" }),
  );
  const frames = out.frames();
  assert.deepEqual(
    frames.map((f) => f.type),
    ["hello", "error"],
  );
  assert.match(log.text(), /not a protocol frame/);
});

test("unknown frame types are ignored", async () => {
  const { out, log } = await converse(
    [noop],
    JSON.stringify({ type: "hello", protocol: PROTOCOL }),
    JSON.stringify({ type: "prefetch" }),
    JSON.stringify({ type: "bye" }),
  );
  assert.equal(out.frames().length, 1);
  assert.match(log.text(), /unknown frame type/);
});

test("an unknown analyzer id is an error frame, not a crash", async () => {
  const { out } = await converse(
    [noop],
    JSON.stringify({ type: "hello", protocol: PROTOCOL }),
    JSON.stringify({ type: "analyze", analyzer: "nope" }),
    JSON.stringify({ type: "bye" }),
  );
  assert.match(out.frames()[1]?.message ?? "", /no analyzer named nope/);
});

test("parseFrame rejects what is not a frame", () => {
  for (const [input, want] of [
    ["not json", /not a protocol frame/],
    ["[1,2,3]", /not a JSON object/],
    ['{"protocol":1}', /no type/],
  ] as const) {
    const parsed = parseFrame(input);
    assert.ok("error" in parsed, `${input} should not parse`);
    assert.match(parsed.error, want);
  }
});

File: pkg/plugin/serve_test.go
// SPDX-License-Identifier: MIT

package plugin

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"io"
	"strings"
	"testing"

	"github.com/adamtait/reviewer/pkg/finding"
)

type fakeHandler struct {
	descriptors []Descriptor
	analyze     func(id string, req AnalyzeRequest) ([]finding.Finding, []string, error)
	calls       []string
}

func (f *fakeHandler) Name() string    { return "fake" }
func (f *fakeHandler) Version() string { return "0.0.1" }
func (f *fakeHandler) Describe() []Descriptor {
	return f.descriptors
}
func (f *fakeHandler) Analyze(_ context.Context, id string, req AnalyzeRequest) ([]finding.Finding, []string, error) {
	f.calls = append(f.calls, id)
	if f.analyze != nil {
		return f.analyze(id, req)
	}
	return nil, nil, nil
}

// converse drives serve with a scripted host and returns the frames it wrote.
func converse(t *testing.T, h Handler, lines ...string) ([]Frame, string, error) {
	t.Helper()
	var out, logs bytes.Buffer
	err := serve(context.Background(), h, strings.NewReader(strings.Join(lines, "\n")+"\n"), &out, &logs)

	var frames []Frame
	r := NewReader(bytes.NewReader(out.Bytes()))
	for {
		f, readErr := r.Read()
		if errors.Is(readErr, ErrClosed) {
			break
		}
		if readErr != nil {
			t.Fatalf("serve wrote something that is not a frame: %v", readErr)
		}
		frames = append(frames, f)
	}
	return frames, logs.String(), err
}

func TestServeFullConversation(t *testing.T) {
	h := &fakeHandler{
		descriptors: []Descriptor{{ID: "demo", Lane: finding.LaneDeterministic, Order: 10, Available: true}},
		analyze: func(id string, req AnalyzeRequest) ([]finding.Finding, []string, error) {
			if req.Root != "/repo" {
				return nil, nil, fmt.Errorf("want the request root, got %q", req.Root)
			}
			return []finding.Finding{{
				RuleID: "demo/found", Lane: finding.LaneDeterministic,
				Confidence: finding.ConfidenceHigh, Severity: finding.SeverityInfo,
				File: "a.ts", Line: 1, Message: "hello",
			}}, []string{"a warning"}, nil
		},
	}
	frames, _, err := converse(t, h,
		`{"type":"hello","protocol":1,"host":"test"}`,
		`{"type":"describe"}`,
		`{"type":"analyze","analyzer":"demo","request":{"root":"/repo"}}`,
		`{"type":"bye"}`,
	)
	if err != nil {
		t.Fatalf("serve returned an error: %v", err)
	}
	got := make([]FrameType, len(frames))
	for i, f := range frames {
		got[i] = f.Type
	}
	want := []FrameType{TypeHello, TypeDescribe, TypeFindings}
	if fmt.Sprint(got) != fmt.Sprint(want) {
		t.Fatalf("want %v, got %v", want, got)
	}
	if frames[0].Protocol != Protocol || frames[0].Plugin != "fake" {
		t.Fatalf("hello frame is wrong: %+v", frames[0])
	}
	if len(frames[2].Findings) != 1 || frames[2].Warnings[0] != "a warning" {
		t.Fatalf("findings frame is wrong: %+v", frames[2])
	}
	if fmt.Sprint(h.calls) != "[demo]" {
		t.Fatalf("want one call to demo, got %v", h.calls)
	}
}

func TestServeRefusesAProtocolMismatch(t *testing.T) {
	frames, _, err := converse(t, &fakeHandler{}, `{"type":"hello","protocol":99}`)
	if err == nil {
		t.Fatal("want an error for an unknown protocol version")
	}
	if !strings.Contains(err.Error(), "protocol 99") {
		t.Fatalf("want the version named, got %v", err)
	}
	if len(frames) != 1 || frames[0].Type != TypeError {
		t.Fatalf("want a single error frame so the host learns why, got %+v", frames)
	}
}

func TestServeRefusesWorkBeforeHello(t *testing.T) {
	frames, _, err := converse(t, &fakeHandler{}, `{"type":"analyze","analyzer":"demo"}`)
	if err == nil || !strings.Contains(err.Error(), "before hello") {
		t.Fatalf("want a handshake-order error, got %v", err)
	}
	if len(frames) != 1 || frames[0].Type != TypeError {
		t.Fatalf("want an error frame, got %+v", frames)
	}
}

// An analyzer that fails must not take the run down (ADR-0013): the host gets an
// error frame for that analyzer and the conversation continues.
func TestServeReportsAnalyzerErrorsAndKeepsGoing(t *testing.T) {
	h := &fakeHandler{
		analyze: func(id string, _ AnalyzeRequest) ([]finding.Finding, []string, error) {
			if id == "broken" {
				return nil, nil, errors.New("the config is missing")
			}
			return nil, nil, nil
		},
	}
	frames, logs, err := converse(t, h,
		`{"type":"hello","protocol":1}`,
		`{"type":"analyze","analyzer":"broken"}`,
		`{"type":"analyze","analyzer":"fine"}`,
		`{"type":"bye"}`,
	)
	if err != nil {
		t.Fatalf("an analyzer error must not fail serve, got %v", err)
	}
	if len(frames) != 3 || frames[1].Type != TypeError || frames[2].Type != TypeFindings {
		t.Fatalf("want hello, error, findings; got %+v", frames)
	}
	if frames[1].Analyzer != "broken" {
		t.Fatalf("the error frame must name the analyzer, got %q", frames[1].Analyzer)
	}
	if !strings.Contains(logs, "the config is missing") {
		t.Fatalf("want the failure on stderr for the run log, got %q", logs)
	}
	if fmt.Sprint(h.calls) != "[broken fine]" {
		t.Fatalf("want both analyzers attempted, got %v", h.calls)
	}
}

func TestServeSurvivesGarbageOnTheWire(t *testing.T) {
	frames, logs, err := converse(t, &fakeHandler{},
		`{"type":"hello","protocol":1}`,
		`oops, a log line leaked onto stdin`,
		`{"type":"bye"}`,
	)
	if err != nil {
		t.Fatalf("a bad line must not fail serve, got %v", err)
	}
	if len(frames) != 2 || frames[1].Type != TypeError {
		t.Fatalf("want hello then an error frame, got %+v", frames)
	}
	if !strings.Contains(logs, "not a protocol frame") {
		t.Fatalf("want the bad line reported on stderr, got %q", logs)
	}
}

func TestServeTreatsAClosedStreamAsANormalExit(t *testing.T) {
	// The host timed out or was killed. There is nothing to clean up and nothing
	// to report, so this is not an error.
	if err := serve(context.Background(), &fakeHandler{}, strings.NewReader(`{"type":"hello","protocol":1}`+"\n"), io.Discard, io.Discard); err != nil {
		t.Fatalf("want a clean exit on a closed stream, got %v", err)
	}
}

func TestServeIgnoresUnknownFrameTypes(t *testing.T) {
	// Additive extensibility: a newer host may send frames this plugin predates.
	frames, logs, err := converse(t, &fakeHandler{},
		`{"type":"hello","protocol":1}`,
		`{"type":"prefetch","analyzer":"demo"}`,
		`{"type":"bye"}`,
	)
	if err != nil {
		t.Fatalf("an unknown frame type must not be fatal, got %v", err)
	}
	if len(frames) != 1 {
		t.Fatalf("want only the hello reply, got %+v", frames)
	}
	if !strings.Contains(logs, `unknown frame type "prefetch"`) {
		t.Fatalf("want the unknown type noted on stderr, got %q", logs)
	}
}
